src/NHibernate/Proxy/DynamicProxy/ProxyCacheEntry.cs

```diff
@@ -43,15 +44,26 @@ public ProxyCacheEntry(System.Type baseType, System.Type[] interfaces)
 		public System.Type BaseType { get; private set; }
 		public System.Type[] Interfaces { get; private set; }
 
+		[NonSerialized]
+		private HashSet<System.Type> _uniqueInterfaces;
+
 		public override bool Equals(object obj)
 		{
 			var that = obj as ProxyCacheEntry;
-			if (ReferenceEquals(null, that))
+			if (ReferenceEquals(null, that) ||
+				hashCode != that.hashCode ||
```

Comment:
I left it as an easy early "exit case". It will most of time speed up non equal cases, while being a very minor overhead for equal cases. I should add a comment for warning the hashcode is only good for early false yielding, not for telling it is equal.